### PR TITLE
Switching to web-executor

### PR DESCRIPTION
**Pull request by VicDeo**
Depends on 
master
- [x] https://github.com/owncloud/core/pull/24957 (merged)
- [x] https://github.com/owncloud/core/pull/25356 (merged)

stable9.1
- [ ] https://github.com/owncloud/core/pull/24957 (was merged in 9.1 before)
- [x] https://github.com/owncloud/core/issues/25366

stable9
- [x] https://github.com/owncloud/core/pull/25222 (merged)
- [x] https://github.com/owncloud/core/pull/25363 (merged)

Right now web executor is used when update is initiated from Web UI

### How to test:
1. Install OC 9.0.3
2. Apply https://github.com/owncloud/core/pull/25222
3. Apply https://github.com/owncloud/core/pull/25356 
4. Apply this PR
5. Specify custom update server url in config.php:    
  ```
'updater.server.url' => 'https://domain.tld/your_update_server/',
```
6. Download master or stable9 daily. 
7. Repack daily applying   
    (for stable9 daily)     
    https://github.com/owncloud/core/pull/25222  https://github.com/owncloud/core/pull/25363    
    
   (for master daily)    
    https://github.com/owncloud/core/pull/25356     
8. Put repacked daily to the location your update server points to

9. Update (from Web UI)

### CODE REVIEW

File: src/Utils/OccRunner.php [hunk 1]
```diff
@@ -32,16 +34,71 @@ class OccRunner {
...
+		}
+		return $decoded;
+	}
+
+	private function runAsRequest($command, $args){
+		$client = new Client();
+		$request = $client->createRequest(
+			'POST',
+			// TODO Get endpoint somehow :-/
+			'http://localhost/~deo/oc-tmp/index.php/occ/' . $command,
```
Comment (DeepDiver1975) [inline src/Utils/OccRunner.php]: bäääh ... just another chicken egg issue ....

Comment (DeepDiver1975) [inline src/Utils/OccRunner.php]: well - we are only in web executor mode is the updater is executed from the browser.
Within the browser we know the address - right?

File: src/Controller/IndexController.php [hunk 2]
```diff
@@ -135,6 +134,9 @@ public function ajaxAction() {
 		$output->setFormatter(new HtmlOutputFormatter($formatter));
 
 		$application->setAutoExit(false);
+
+		$endpoint = preg_replace('/(updater\/|updater\/index.php)$/', '', $this->request->server('REQUEST_URI'));
```
Comment (DeepDiver1975) [inline src/Controller/IndexController.php]: I'm honestly bad with regex :see_no_evil: 

Will this work for setups where index.php is not part of the url?

THX

Comment (VicDeo) [inline src/Controller/IndexController.php]: @DeepDiver1975 yes. It stands for 
(updater/`OR`updater/index.php)`EOL`

**Comment by DeepDiver1975** [comment]
here we go ... my long standing cry for unit test ... please

**Comment by DeepDiver1975** [comment]
Unit tests please

**Comment by VicDeo** [comment]
@DeepDiver1975 tests added

**Comment by VicDeo** [comment]
@PVince81  @DeepDiver1975 not a WIP anymore, please review

**Comment by PVince81** [comment]
I tested with the steps above, going from a 9.0.3 tarball + patches to stable9+patches and then to 9.1.0RC1 tarball + patches.

Also I observed the access log of Apache and saw that "occ" was used over the web executor.

:+1: works

**Comment by PVince81** [comment]
I suggest we merge this and let @davitol test with tomorrow's stable9/master dailies for easier testing.

Would still require patching of 9.0.3.

File: src/Tests/Http/RequestTest.php [hunk 3]
```diff
@@ -23,4 +23,63 @@ public function testPostParameter($vars, $key, $expected){
...
+	}
+
+
+	public function hostProvider(){
+		return [
+			[ [ 'headers'=> [ 'SERVER_NAME' => 'jump' ] ], 'jump', null ],
+			[ [ 'headers'=> [ 'HTTP_HOST'=> 'duck', 'SERVER_NAME' => 'jump'] ], 'duck' ],
+			[ [ 'headers'=> [ 'HTTP_X_FORWARDED_HOST'=>'go', 'HTTP_HOST'=> 'duck', 'SERVER_NAME' => 'jump'] ], 'go' ],
+			[ [ 'headers'=> [ 'HTTP_X_FORWARDED_HOST'=>'go,', 'HTTP_HOST'=> 'duck', 'SERVER_NAME' => 'jump'] ], 'go' ],
+			[ [ 'headers'=> [ 'HTTP_X_FORWARDED_HOST'=>'run,forrest,run', 'HTTP_HOST'=> 'duck', 'SERVER_NAME' => 'jump'] ], 'run' ],
```
Comment (PVince81) [inline src/Tests/Http/RequestTest.php]: :laughing:

File: src/Utils/OccRunner.php [hunk 4]
```diff
@@ -32,16 +35,84 @@ class OccRunner {
...
+			}
+			$response = $this->runAsRequest($command, $args);
+			$decodedResponse = json_decode($response, true);
+			return $decodedResponse['response'];
+		}
+	}
+
+	public function runJson($command, $args = []){
+		$plain = $this->run($command, $args, true);
+		preg_match_all('!(\{.*\})!ms', $plain, $matches);
```
Comment (PVince81) [inline src/Utils/OccRunner.php]: add a comment, what is this about ?

**Comment by PVince81** [comment]
Code looks fine, too. Great job and thanks for the detailed test steps !

**Comment by PVince81** [comment]
This will need to be backported to stable9 for 9.0.4.

**Comment by PVince81** [comment]
@DeepDiver1975 @georgehrke second review please

**Comment by georgehrke** [comment]
Looks good except for the missing explanation of the magic regular expression

**Comment by PVince81** [comment]
Great! @georgehrke agreed ?

**Comment by VicDeo** [comment]
@PVince81 please hold on for a minute, I'm squashing history

**Comment by VicDeo** [comment]
@PVince81 squashed

**Comment by VicDeo** [comment]
Stable9.1: https://github.com/owncloud/updater/pull/362
Stable9: https://github.com/owncloud/updater/pull/363

**Comment by georgehrke** [comment]
👍

**Comment by PVince81** [comment]
QA ticket to redo a full test once the backports are merged: https://github.com/owncloud/QA/issues/277

**Pull request merged**